### ISSUE TITLE
[Phase 3] Supervised Model with Custom Loss

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/fraud-detection-adaptive#5]
## Summary

Implement XGBoost classifier with cost-sensitive custom objective function. This phase creates the supervised scoring layer that learns from labeled data while accounting for asymmetric fraud costs.

**Parent Issue**: #1

## Objectives

- Define cost matrix encoding FN >> FP penalty
- Implement custom gradient/hessian for XGBoost
- Create training wrapper with proper class handling
- Enable cross-validation with stratified folds

## Dependencies

- Phase 0 must be complete (synthetic data generator)
- Phase 1 must be complete (feature pipeline)

## Task Checklist

- [ ] Implement `src/models/cost_matrix.py` - cost matrix definition
- [ ] Implement `src/models/custom_loss.py` - XGBoost custom objective
- [ ] Implement `src/models/xgboost_classifier.py` - training wrapper
- [ ] Create `tests/test_supervised.py` - unit tests
- [ ] Validate custom loss with known inputs

### Cost Matrix

📄 **File:** `src/models/cost_matrix.py`

```python
"""Cost matrix for fraud detection."""

from dataclasses import dataclass
from typing import Tuple
import numpy as np


@dataclass
class CostMatrix:
    """
    Cost matrix encoding business impact of prediction errors.

    In fraud detection:
    - False Negative (FN): Fraud not caught = direct financial loss
    - False Positive (FP): Legitimate blocked = customer friction

    Typical ratio: FN costs 10-100x more than FP.
    """

    fn_cost: float = 100.0  # Cost of missing a fraud (False Negative)
    fp_cost: float = 1.0    # Cost of blocking legitimate (False Positive)
    tn_cost: float = 0.0    # True Negative: correct approval
    tp_cost: float = 0.0    # True Positive: correct fraud catch

    def get_sample_weights(self, y_true: np.ndarray) -> np.ndarray:
        """
        Convert cost matrix to per-sample weights.

        For training, we weight positive samples (fraud) by FN cost
        and negative samples by FP cost, normalized.
        """
        weights = np.where(y_true == 1, self.fn_cost, self.fp_cost)
        # Normalize to mean of 1
        return weights / weights.mean()

    def calculate_cost(
        self,
        y_true: np.ndarray,
        y_pred: np.ndarray
    ) -> float:
        """Calculate total cost given predictions."""
        fn = ((y_true == 1) & (y_pred == 0)).sum()
        fp = ((y_true == 0) & (y_pred == 1)).sum()
        tn = ((y_true == 0) & (y_pred == 0)).sum()
        tp = ((y_true == 1) & (y_pred == 1)).sum()

        return (
            fn * self.fn_cost +
            fp * self.fp_cost +
            tn * self.tn_cost +
            tp * self.tp_cost
        )

    def cost_ratio(self) -> float:
        """Return FN/FP cost ratio."""
        return self.fn_cost / self.fp_cost
```

### Custom Loss Function

📄 **File:** `src/models/custom_loss.py`

```python
"""Custom XGBoost objective for cost-sensitive fraud detection."""

import numpy as np
from typing import Tuple


def weighted_binary_cross_entropy(
    y_true: np.ndarray,
    y_pred: np.ndarray,
    fn_cost: float = 100.0,
    fp_cost: float = 1.0
) -> Tuple[np.ndarray, np.ndarray]:
    """
    Weighted binary cross-entropy loss for XGBoost.

    XGBoost requires gradient (first derivative) and hessian (second derivative)
    of the loss function with respect to the raw predictions (before sigmoid).

    The loss is:
        L = -w * [y * log(p) + (1-y) * log(1-p)]

    where:
        - p = sigmoid(raw_pred)
        - w = fn_cost if y=1 else fp_cost

    Args:
        y_true: True labels (0 or 1)
        y_pred: Raw predictions (before sigmoid)
        fn_cost: Cost multiplier for false negatives
        fp_cost: Cost multiplier for false positives

    Returns:
        Tuple of (gradient, hessian) arrays
    """
    # Clip predictions to avoid overflow in exp
    y_pred = np.clip(y_pred, -500, 500)

    # Sigmoid transformation
    prob = 1.0 / (1.0 + np.exp(-y_pred))

    # Clip probabilities to avoid log(0)
    eps = 1e-7
    prob = np.clip(prob, eps, 1 - eps)

    # Sample weights based on true label
    weights = np.where(y_true == 1, fn_cost, fp_cost)

    # Gradient: w * (p - y)
    grad = weights * (prob - y_true)

    # Hessian: w * p * (1 - p)
    hess = weights * prob * (1 - prob)

    # Ensure hessian is positive (required for XGBoost convergence)
    hess = np.maximum(hess, 1e-7)

    return grad, hess


def make_cost_sensitive_objective(
    fn_cost: float = 100.0,
    fp_cost: float = 1.0
):
    """
    Factory function to create XGBoost custom objective.

    Usage:
        objective = make_cost_sensitive_objective(fn_cost=100, fp_cost=1)
        model = xgb.XGBClassifier(objective=objective)
    """
    def objective(y_true: np.ndarray, y_pred: np.ndarray):
        return weighted_binary_cross_entropy(y_true, y_pred, fn_cost, fp_cost)

    return objective


def cost_sensitive_eval_metric(
    fn_cost: float = 100.0,
    fp_cost: float = 1.0
):
    """
    Custom evaluation metric for monitoring during training.

    Returns tuple of (metric_name, metric_value, higher_is_better)
    """
    def eval_metric(y_pred: np.ndarray, dtrain) -> Tuple[str, float]:
        y_true = dtrain.get_label()

        # Convert raw predictions to probabilities
        prob = 1.0 / (1.0 + np.exp(-y_pred))

        # Use 0.5 threshold for evaluation
        y_pred_binary = (prob >= 0.5).astype(int)

        # Calculate weighted cost
        fn = ((y_true == 1) & (y_pred_binary == 0)).sum()
        fp = ((y_true == 0) & (y_pred_binary == 1)).sum()

        cost = fn * fn_cost + fp * fp_cost

        # Normalize by total positive weight
        total_weight = y_true.sum() * fn_cost + (1 - y_true).sum() * fp_cost
        normalized_cost = cost / total_weight if total_weight > 0 else 0

        return 'weighted_cost', normalized_cost

    return eval_metric
```

### XGBoost Classifier Wrapper

📄 **File:** `src/models/xgboost_classifier.py`

```python
"""XGBoost classifier with cost-sensitive training."""

import xgboost as xgb
import numpy as np
from typing import Optional, Dict, Any
from sklearn.model_selection import StratifiedKFold

from src.models.cost_matrix import CostMatrix
from src.models.custom_loss import make_cost_sensitive_objective


class CostSensitiveXGBoost:
    """
    XGBoost classifier with cost-sensitive custom objective.

    Handles extreme class imbalance through:
    1. Custom loss function weighting FN >> FP
    2. scale_pos_weight for additional balancing
    3. Stratified cross-validation
    """

    def __init__(
        self,
        cost_matrix: Optional[CostMatrix] = None,
        n_estimators: int = 100,
        max_depth: int = 6,
        learning_rate: float = 0.1,
        subsample: float = 0.8,
        colsample_bytree: float = 0.8,
        random_state: int = 42,
        use_custom_objective: bool = True
    ):
        """
        Args:
            cost_matrix: Cost matrix for weighting errors
            n_estimators: Number of boosting rounds
            max_depth: Maximum tree depth
            learning_rate: Step size shrinkage
            subsample: Fraction of samples for each tree
            colsample_bytree: Fraction of features for each tree
            random_state: Random seed
            use_custom_objective: Whether to use custom loss (vs class_weight)
        """
        self.cost_matrix = cost_matrix or CostMatrix()
        self.use_custom_objective = use_custom_objective
        self.random_state = random_state

        self.params = {
            'n_estimators': n_estimators,
            'max_depth': max_depth,
            'learning_rate': learning_rate,
            'subsample': subsample,
            'colsample_bytree': colsample_bytree,
            'random_state': random_state,
            'eval_metric': 'auc',
            'use_label_encoder': False,
        }

        self.model: Optional[xgb.XGBClassifier] = None
        self._is_fitted = False

    def fit(
        self,
        X: np.ndarray,
        y: np.ndarray,
        eval_set: Optional[list] = None,
        verbose: bool = False
    ) -> 'CostSensitiveXGBoost':
        """
        Fit the model on training data.

        Args:
            X: Feature matrix
            y: Binary labels (0 or 1)
            eval_set: List of (X, y) tuples for evaluation
            verbose: Whether to print training progress
        """
        # Calculate class ratio for scale_pos_weight
        n_neg = (y == 0).sum()
        n_pos = (y == 1).sum()
        scale_pos_weight = n_neg / n_pos if n_pos > 0 else 1.0

        if self.use_custom_objective:
            objective = make_cost_sensitive_objective(
                fn_cost=self.cost_matrix.fn_cost,
                fp_cost=self.cost_matrix.fp_cost
            )
            self.model = xgb.XGBClassifier(
                objective=objective,
                **self.params
            )
        else:
            # Fall back to scale_pos_weight if custom objective fails
            self.model = xgb.XGBClassifier(
                scale_pos_weight=scale_pos_weight * self.cost_matrix.cost_ratio(),
                **self.params
            )

        self.model.fit(
            X, y,
            eval_set=eval_set,
            verbose=verbose
        )

        self._is_fitted = True
        return self

    def predict_proba(self, X: np.ndarray) -> np.ndarray:
        """Return probability predictions."""
        if not self._is_fitted:
            raise ValueError("Model must be fitted before prediction")
        return self.model.predict_proba(X)

    def predict(self, X: np.ndarray, threshold: float = 0.5) -> np.ndarray:
        """Return binary predictions at given threshold."""
        proba = self.predict_proba(X)[:, 1]
        return (proba >= threshold).astype(int)

    def score(self, X: np.ndarray) -> np.ndarray:
        """Return fraud probability scores."""
        return self.predict_proba(X)[:, 1]

    def cross_validate(
        self,
        X: np.ndarray,
        y: np.ndarray,
        n_splits: int = 5
    ) -> Dict[str, np.ndarray]:
        """
        Perform stratified cross-validation.

        Returns dict with 'scores' and 'costs' per fold.
        """
        skf = StratifiedKFold(n_splits=n_splits, shuffle=True, random_state=self.random_state)

        scores = []
        costs = []

        for train_idx, val_idx in skf.split(X, y):
            X_train, X_val = X[train_idx], X[val_idx]
            y_train, y_val = y[train_idx], y[val_idx]

            # Clone and fit
            fold_model = CostSensitiveXGBoost(
                cost_matrix=self.cost_matrix,
                use_custom_objective=self.use_custom_objective,
                random_state=self.random_state,
                **{k: v for k, v in self.params.items() if k != 'random_state'}
            )
            fold_model.fit(X_train, y_train)

            # Evaluate
            y_pred = fold_model.predict(X_val)
            cost = self.cost_matrix.calculate_cost(y_val, y_pred)
            costs.append(cost)

            # ROC AUC
            from sklearn.metrics import roc_auc_score
            try:
                auc = roc_auc_score(y_val, fold_model.score(X_val))
            except ValueError:
                auc = 0.5
            scores.append(auc)

        return {
            'auc_scores': np.array(scores),
            'costs': np.array(costs)
        }
```

## Verification Commands

```bash
# Run supervised model tests
pytest tests/test_supervised.py -v

# Test custom loss correctness
python -c "
from src.models.custom_loss import weighted_binary_cross_entropy
import numpy as np

# Test: gradient should be positive when pred > true
y_true = np.array([0, 0, 1, 1])
y_pred = np.array([2, -2, 2, -2])  # Raw predictions

grad, hess = weighted_binary_cross_entropy(y_true, y_pred)
print(f'Gradients: {grad}')
print(f'Hessians: {hess}')
print(f'Hessians all positive: {(hess > 0).all()}')
"
```

## Definition of Done

- [ ] Model achieves >0.5 recall on fraud class
- [ ] Precision-recall AUC > baseline (0.5)
- [ ] Custom loss correctly penalizes FN >> FP
- [ ] Cross-validation runs without errors
- [ ] Fallback to class_weight works when custom objective fails
- [ ] All tests pass

## Debugging Scenarios

| Symptom | Likely Cause | Fix |
|---------|--------------|-----|
| All predictions are 0 | FN cost too low | Increase fn_cost, check scale_pos_weight |
| All predictions are 1 | FP cost too low | Increase fp_cost |
| NaN gradients | log(0) in loss | Add epsilon clipping |
| No convergence | Learning rate too high | Reduce learning_rate to 0.01 |

## Files to Modify

| File | Lines | Action | Description |
|------|-------|--------|-------------|
| `src/models/cost_matrix.py` | - | Create | Cost matrix class |
| `src/models/custom_loss.py` | - | Create | XGBoost custom objective |
| `src/models/xgboost_classifier.py` | - | Create | Training wrapper |
| `tests/test_supervised.py` | - | Create | Unit tests |

## Concepts to Understand

Before implementing, review:
- **XGBoost objective functions**: Must return gradient and hessian
- **Sigmoid function**: `1 / (1 + exp(-x))` for probability
- **Cross-entropy loss**: `-y*log(p) - (1-y)*log(1-p)`
- **Stratified K-Fold**: Preserves class ratio in each fold
